src/engine/Sim3D.ts

```diff
@@ -224,21 +233,59 @@ export class Sim3D {
     return obj.object;
   }
 
-  addGameObject(spec: SimObjectSpec): ISimObjectRef | undefined {
+  private addGameObject<T1>(
+    spec: SimObjectSpec,
+    typeT: { new (simObject: SimObject, rootObject: SimObject): T1 }
+  ): T1 | undefined {
     const obj = this.objectFactories.makeObject(spec);
     if (obj === undefined) {
-      return obj;
+      return undefined;
     }
+    let body = this.world.createBody(obj.getBodySpecs());
+    obj.setBody(body);
+    body.createFixture(obj.getFixtureDef());
 
-    obj.addToScene();
+    this.addToScene(obj);
     this.simObjects.set(obj.guid, {
       type: obj.type,
       object: obj,
     });
 
-    return {
+    let simObjectRef = {
       guid: obj.guid,
       type: obj.type,
     };
+
+    let rootObject = (this.getSimObject(simObjectRef) as unknown) as SimObject;
+    if (!rootObject) {
+      throw new Error(
+        `Unable to get SimObject with guid "${simObjectRef.guid}"`
+      );
+    }
+
+    let handle = new typeT(obj, rootObject);
+    return handle;
+  }
+
+  addBall(spec: IBallSpec): BallHandle | undefined {
```

Comment:
See my overall comment. We could combine the `addBall` and `addBox` methods into a `addBasicShape` method that takes in a shape spec.